### PR TITLE
feat(lsp): quick-fix code actions for lint diagnostics

### LINKED ISSUE #776: feat(lsp): quick-fix code actions for lint diagnostics
> **Assignment:** Dev B (independent-features stream) | **Phase 2**
> **Blocked by:** #766 (cache), #767 (seam) | **Blocks:** #779 (reuses this handler)
> **Sync point:** none for start -- needs only #766/#767, NOT the cursor resolver (#768).
> Dev B also owns the additive `lint.QuickFixFor` gap-fill.

**PR unit:** R5 | **Complexity:** M/L | **Depends on:** #766 (cache), #767 (seam)
**Not** dependent on the cursor resolver (#768) -- uses diagnostic ranges.

## Scope

Implement `textDocument/codeAction` (QuickFix kind) that turns existing lint
diagnostics into one-click fixes.

- We already publish diagnostics with `Code` = rule name
  (`pkg/lsp/diagnostics.go:89`). The incoming
  `params.Context.Diagnostics` carry that rule name.
- Map rule -> `WorkspaceEdit`:
  - **deprecated-field** -> replace the old key with its
    `deprecatedReplacement` (e.g. `onError` -> `continueOnError`, with value
    translation `continue`->`true` / `fail`->`false`).
  - **redundant-depends-on** -> remove the redundant `dependsOn` entry/list.
  - **unused-resolver** -> remove the unused resolver block.
- Register `codeActionFeature()` via the seam; advertise
  `CodeActionProvider{CodeActionKinds:[QuickFix]}`.

### Gap (additive, `pkg/lint`)

The deprecated replacement map is private (`lookupDeprecatedField`,
`pkg/lint/deprecated.go:33`). Export a fix builder so `pkg/lsp` doesn't
re-implement tag reading and stays in lockstep with lint:

~~~go
// QuickFixFor returns edits that resolve the finding when a deterministic fix
// exists (deprecated-field replacement, redundant dependsOn, unused resolver).
func QuickFixFor(sol *solution.Solution, f *Finding) (edits []refactor.TextEdit, ok bool)
~~~

Use `pkg/refactor`'s text-edit/apply primitives to build edits.

## Files

- New: `pkg/lsp/codeaction.go` (+ test)
- New/changed: `pkg/lint` (`QuickFixFor`) (+ test)
- Changed: `pkg/lsp/server.go` (one line: `codeActionFeature()`)

## Tests / docs

- Table tests per fixable rule producing the correct edit.
- Integration test via `runLSPSession` (open a doc with a deprecated field ->
  code action returns the replacement edit).
- Tutorial + showcase (include one deprecated-field example to demo the fix).

## Acceptance criteria

- [ ] Each supported diagnostic offers a working quick fix.
- [ ] `lint.QuickFixFor` is the single source of fix logic (no duplication in lsp).
- [ ] `CodeActionProvider` advertised via seam.

### PR DESCRIPTION
**Pull request by kcloutie**
## Summary

Implements `textDocument/codeAction` (QuickFix) that turns existing lint diagnostics into one-click editor fixes. The fix logic lives **once** in `pkg/lint` (`QuickFixFor`), so the applied edit always matches what the linter recommends -- no duplication in the LSP layer.

## Changes

- **New `pkg/lint/quickfix.go`** -- `QuickFixFor(sol, f, registry) ([]refactor.TextEdit, ok)`:
  - **deprecated-field**: replace `onError` with its tagged replacement `continueOnError`, translating `continue` -> `true` / `fail` -> `false`. The replacement name is read from the struct tag via lint's existing `lookupDeprecatedField`, so it tracks lint.
  - **redundant-depends-on**: recomputes the redundant set from the solution (not the message text) using the **same provider-aware dependency lookup the finding used**, so the removal stays in lockstep with the finding even when a provider's custom `ExtractDependencies` diverges from generic extraction; removes the whole `dependsOn:` entry when all entries are redundant, else just the redundant list elements.
  - **unused-resolver**: removes the resolver block. When it is the **sole** resolver, removes the parent `resolvers:` key instead, so no null-valued key is left behind (which would turn the warning into a schema error).
  - Guarded throughout: any odd input yields `(nil, false)`, never a broken edit.
- **New `pkg/refactor/edits.go`** -- exported, source-preserving edit builders `RemoveMappingEntry`, `RemoveSequenceElement`, `ReplaceMappingKeyAndValue`, reusing the existing byte-edit/`Apply` model and the indentation-scan span logic from `ExtractCall` (no duplication; `extract.go` untouched). No import cycle (`pkg/refactor` does not depend on `pkg/lint`).
- **New `pkg/lsp/codeaction.go`** -- `codeActionFeature()` registered via the #767 seam; advertises `CodeActionProvider{CodeActionKinds:[QuickFix]}`. The handler re-lints the cached solution (#766), keeps findings that are fixable and relevant to the request range/diagnostics (respecting `Context.Only`), and returns each as a QuickFix carrying the `WorkspaceEdit`. It passes the server's provider registry into `QuickFixFor` so the applied edit uses the identical dependency extraction the diagnostics were computed with.

## Tests / docs

- Per-rule fix tests -- every applied output **re-parses and passes `ValidateSpec`**; edit-builder table tests; handler tests (Only filter, advertise, registration); a **sole-resolver regression** (guards against the warning->error failure mode); a **provider-lockstep regression** (a provider whose custom `ExtractDependencies` diverges from generic must not cause removal of a non-redundant `dependsOn` entry); and an integration test via the `runLSPSession` harness (#769) asserting the deprecated-field quick fix and advertised capability.
- `docs/tutorials/lsp-tutorial.md` documents the quick fixes.
- `go build/vet`, `go test ./pkg/refactor/... ./pkg/lint/... ./pkg/lsp/... ./tests/integration/...`, `task lint:changed` (0 new issues) -- all pass.
- A dedicated code-review pass found a file-degrading edge case (sole-resolver removal producing a schema-error document); it was fixed + regression-tested before opening.

## Acceptance criteria

- [x] Each supported diagnostic offers a working quick fix
- [x] `lint.QuickFixFor` is the single source of fix logic (no duplication in lsp)
- [x] `CodeActionProvider` advertised via the seam

Depends on #766 (cache) and #767 (seam), both merged; not dependent on #768. This handler is reused by #779.

Closes #776

### CODE REVIEW

File: pkg/lint/quickfix.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+	res, ok := sol.Spec.Resolvers[name]
+	if !ok || res == nil || len(res.DependsOn) == 0 {
+		return nil, false
+	}
+
+	// Recompute the redundant set from the solution. A nil lookup performs only
+	// generic ValueRef-based extraction; the lint rule additionally consults the
+	// registry's per-provider ExtractDependencies. For all built-in providers the
+	// two converge, so the redundant set matches the finding. A plugin provider
+	// with custom dependency extraction could diverge; if any element cannot be
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/lint/quickfix.go]: Good catch -- confirmed with a reproduction and fixed in 6b53632.

You are right that a provider's `ExtractDependencies` *replaces* (not augments) generic input extraction, so the nil-lookup recompute could diverge from the registry-aware finding. I reproduced it: a provider that reports only a subset of the inferred deps made the finding flag a single redundant entry, while `QuickFixFor` (nil) flagged *all* of them and therefore removed the whole `dependsOn:` block -- deleting a non-redundant ordering dependency the finding never flagged. So the `guards below decline the fix` comment was not actually guaranteed.

Fix: `QuickFixFor` now takes the provider registry, and `quickFixRedundantDependsOn` recomputes the redundant set with the same `DescriptorLookup` that `lintRedundantDependsOn` uses -- so the removal is in exact lockstep with the finding regardless of provider-specific extraction. Added a regression test (`TestQuickFixFor_RedundantDependsOn_ProviderLookupLockstep`) using a provider whose custom extraction diverges from generic, and corrected the misleading comment. Resolving.

File: pkg/lsp/codeaction.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+// lineInRange reports whether a 0-based line falls within an LSP range's line
+// span (inclusive).
+func lineInRange(line uint32, r protocol.Range) bool {
+	return line >= r.Start.Line && line <= r.End.Line
+}
+
+// rangesOverlapLines reports whether two LSP ranges share any line.
+func rangesOverlapLines(a, b protocol.Range) bool {
+	return a.Start.Line <= b.End.Line && b.Start.Line <= a.End.Line
+}
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/lsp/codeaction.go:144]: Thanks -- you are right that LSP ranges are half-open, so a full-line selection ending at `{nextLine, 0}` makes these helpers treat that next line as included.

I am treating this as a non-blocking nit rather than changing it. The line-level overlap in `findingMatchesRequest` is intentionally coarse (see its doc comment) to stay robust to the column differences between a finding's squiggle and the client's cursor/selection. The only consequence of the inclusive end is that a valid quick fix may be *offered* on one adjacent line at a selection boundary -- it never changes the edit itself (which is always the correct fix for its own finding), and it never attaches to the wrong diagnostic (`matchingDiagnostics` keys off the rule name, not the range). Since the worst case is over-offering a correct action by a single line, I would rather keep the coarse matching than risk under-offering fixes at boundaries by tightening the end to exclusive. Resolving as a non-blocking nit.

**Comment by kcloutie** [comment]
Addressed the Copilot review finding (matchingDiagnostics attaching all same-rule diagnostics to every action) in aaef0ab7: matchingDiagnostics now filters attached diagnostics by rule name AND overlap with the finding's own line (falling back to the request range when the finding has no resolved position), so each quick fix advertises only the diagnostic it actually resolves. Added a regression test (TestCodeAction_AttachesOnlyItsOwnDiagnostic) with two deprecated-field diagnostics on different lines.

**Comment by kcloutie** [comment]
Addressed the Copilot finding on `ReplaceMappingKeyAndValue` (span computed from decoded `yaml.Node.Value` length could be wrong for quoted scalars with escapes) in 45ee5582: `scalarTokenEnd` now computes the token's end by scanning the SOURCE bytes -- handling double-quoted `\"` and single-quoted `''` escapes -- and declines (returns an error, no edit) for unterminated, block/folded, or multiline plain scalars. Added escaped-quote regression tests (`TestReplaceMappingKeyAndValue_EscapedQuotedValue`). Also merged origin/main to resolve the conflict from #794 (schema-driven completion) landing.

File: pkg/refactor/edits.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+// simple single/double-quoted scalars without escapes (the onError values
+// "continue"/"fail" are always in this class); block/folded scalars are not a
+// target of the replace helper and fall back to the decoded length.
+func scalarRawLen(n *yaml.Node) int {
+	l := len(n.Value)
+	if n.Style&(yaml.DoubleQuotedStyle|yaml.SingleQuotedStyle) != 0 {
+		l += 2 // opening and closing quote
+	}
+	return l
+}
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/refactor/edits.go:253]: Fixed in 45ee5582. ReplaceMappingKeyAndValue no longer derives the span from the decoded n.Value length (scalarRawLen is removed). scalarTokenEnd now computes the token end by scanning the raw source bytes: for double-quoted scalars it skips backslash-escaped bytes and stops at the true closing quote; for single-quoted it handles the '' escape; a real newline inside the quotes (a folded multi-line scalar), an unterminated quote, a multi-line plain scalar, or a block/folded scalar are declined (error, no edit) rather than mis-spanned. Added TestReplaceMappingKeyAndValue_EscapedQuotedValue covering the double- and single-quoted escape cases.

**Comment by kcloutie** [comment]
Reviewed both suppressed comments from the latest Copilot review:

**edits.go:113 (sequence-marker guard) -- addressed in ee266690.** The guard checked only for a leading `-` while the doc/error say `- `. In practice `--foo` was never reachable (`RemoveSequenceElement` is only ever called with a `NodeMap` `[i]` path, which YAML only produces for a genuine block sequence element), so this was not a live bug -- but the exported helper's guard should match its documented contract, so I factored it into `isSequenceMarker`, which requires `-` followed by whitespace or end-of-line (accepts `- value` and a bare `-` with a nested value; rejects `--foo`/`-foo`). Added a unit test.

**quickfix.go:214 (N YAML parses in the partial-dependsOn loop) -- acknowledged, not changing.** This path only runs for the partial-redundant-dependsOn quick fix and re-parses one small in-memory solution a handful of times (once per redundant entry) per code-action request. Removing it would mean threading a shared parse context through the exported edit builders, adding API complexity for no meaningful latency win. Keeping the simpler, self-contained helpers is the intended tradeoff.
